src/riak_kv_ddl_compiler.erl

```diff
@@ -0,0 +1,1458 @@
+-define(POSNOSTART,  1).
+-define(LINENOSTART, 1).
+-define(ARITYOF1,    1).
+
+-type expr()   :: tuple().
+-type exprs()  :: [expr()].
+-type guards() :: [exprs()].
+-type ast()    :: [expr() | exprs() | guards()].
+%% maps() are an aggregration of [map()]
+-type maps()   :: {maps, [[#riak_field_v1{}]]}.
+-type map()    :: {map,  [#riak_field_v1{}]}.
+
+-spec make_helper_mod(#ddl_v1{}) -> {module, atom()} | {'error', tuple()}.
+make_helper_mod(#ddl_v1{} = DDL) ->
+    make_helper_mod(DDL, "/tmp", ?NODEBUGOUTPUT).
+
+-spec make_helper_mod(#ddl_v1{}, string()) ->
+			     {module, atom()} | {'error', tuple()}.
+make_helper_mod(#ddl_v1{} = DDL, OutputDir) ->
+    make_helper_mod(DDL, OutputDir, ?DEBUGOUTPUT).
+
+-spec make_helper_mod(#ddl_v1{}, string(), boolean()) ->
+			     {module, atom()} | {'error', tuple()}.
+make_helper_mod(#ddl_v1{} = DDL, OutputDir, HasDebugOutput) ->
+    case validate_ddl(DDL) of
+	true ->
+	    {Module, AST} = compile(DDL, OutputDir, HasDebugOutput),
+	    {ok, Module, Bin} = compile:forms(AST),
+	    SpoofBeam = "/tmp/" ++ atom_to_list(Module) ++ ".beam",
+	    {module, Module} = code:load_binary(Module, SpoofBeam, Bin);
+	{false, Errs} ->
+	    {error, {invalid_ddl, Errs}}
+    end.
+
+%%
+%% Funs to validate id the DDL definition is correct
+%%
+
+validate_ddl(#ddl_v1{fields = Fields}) ->
+    Ret1 = validate_fields(Fields, []),
+    Ret2 = validate_positions(Fields),
+    case lists:merge(Ret1, Ret2) of
+	[]   -> true;
+	Errs -> {false, Errs}
+    end.
+
+validate_positions(Fields) ->
+    SortFn = fun(#riak_field_v1{position = A}, #riak_field_v1{position = B}) ->
+		     if
+			 A =< B -> true;
+			 A >  B -> false
+		     end
+	     end,
+    Fs2 = lists:sort(SortFn, Fields),
+    validate_pos2(Fs2, ?POSNOSTART, []).
+
+validate_pos2([], _NPos, Acc) ->
+    lists:flatten(lists:reverse(Acc));
+validate_pos2([#riak_field_v1{type     = {map, MapFields},
+			      position = NPos} | T], NPos, Acc) ->
+    NewAcc = case validate_positions(MapFields) of
+		 [] -> Acc;
+		 A  -> [A | Acc]
+	     end,
+    validate_pos2(T, NPos + 1, NewAcc);
+validate_pos2([#riak_field_v1{position = NPos} | T], NPos, Acc) ->
+    validate_pos2(T, NPos + 1, Acc);
+validate_pos2([H | T], NPos, Acc) ->
+    validate_pos2(T, NPos + 1, [{wrong_position, {expected, NPos}, H} | Acc]).
+
+validate_fields([], []) ->
+    [];
+validate_fields([], Acc) ->
+    lists:flatten(lists:reverse(Acc));
+validate_fields([#riak_field_v1{name     = N,
+				position = P,
+				type     = Ty,
+				optional = O} | T], Acc)
+  when is_list(N),             % guard 1
+       is_integer(P)           % guard 2
+       andalso P > 0,
+       Ty == binary            % guard 3
+       orelse Ty =:= integer
+       orelse Ty =:= float
+       orelse Ty =:= boolean
+       orelse Ty =:= set
+       orelse Ty =:= timestamp
+       orelse Ty =:= any,
+       is_boolean(O) ->        % guard 4
+    validate_fields(T, Acc);
```

Comment:
It is hard to tell what this does, brackets around the guards would make this clearer.